Random failures in tailing logs while creating workloads

### Please fill out the issue checklist below and provide ALL the requested information.

- [x] I reviewed open and closed Github issues that may be related to my problem.
- [x] I am reporting a bug that others will be able to reproduce.

### Describe the bug
While creating workloads from local source using `workload apply` command and using the `--tail` flag to follow the progress, I have run into errors which causes the tail to stop with the following error message.

```
unexpected error: read tcp 192.168.1.14:55499->34.205.136.22:443: use of closed network connection
panic: interface conversion: interface is nil, not client.Object

goroutine 489 [running]:
github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime/wait.UntilCondition({0x103369be8, 0xc0008aa480}, {0x10337bfe0, 0xc000130a08}, {{0xc0013a38e4?, 0xc0007ecf80?}, {0xc0013a38e0?, 0xc00036d5c0?}}, {0x103379798, 0xc0007cab60}, ...)
        /Users/adhol/projects/apps-cli-plugin/pkg/cli-runtime/wait/wait.go:44 +0x2b1
github.com/vmware-tanzu/apps-cli-plugin/pkg/commands.(*WorkloadOptions).WaitToBeReady.func1({0x103369be8, 0xc0008aa480})
        /Users/adhol/projects/apps-cli-plugin/pkg/commands/workload.go:780 +0xa5
github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime/wait.Race.func2(0x0?)
        /Users/adhol/projects/apps-cli-plugin/pkg/cli-runtime/wait/wait.go:100 +0x77
created by github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime/wait.Race
        /Users/adhol/projects/apps-cli-plugin/pkg/cli-runtime/wait/wait.go:97 +0x133
Error: exit status 2

✖  exit status 2 
```

### Expected behavior
Apps CLI should continue to tail logs and handle this error gracefully.

### Steps to Reproduce
Create a workload using the --tail flag as follows:
```
tanzu apps workload create tanzu-java-web-app-dockerfile \
  --namespace dev \
  --git-branch main \
  --git-repo https://github.com/garethjevans/tanzu-java-web-app-dockerfile \
  --label apps.tanzu.vmware.com/has-tests=true \
  --label app.kubernetes.io/part-of=test-workload \
  --type web \
  --param dockerfile=./Dockerfile \
  --yes --tail
```

### Screenshots
![Screen Shot 2023-02-14 at 3 29 14 PM](https://user-images.githubusercontent.com/7171846/219467713-b7588720-456b-4c7b-a5c0-0b6ca1332f32.png)


### Version (Apps plugin version, Version of K8s running on cluster)
Put the output of the following commad
```sh
tanzu version && tanzu apps version
``` 
![Screen Shot 2023-02-16 at 2 30 09 PM](https://user-images.githubusercontent.com/7171846/219467782-bbc5bb98-638c-407f-ab05-eba73ed9ad9d.png)

--- a/pkg/cli-runtime/wait/wait.go
+++ b/pkg/cli-runtime/wait/wait.go
@@ -41,16 +41,21 @@ func UntilCondition(ctx context.Context, watchClient client.WithWatch, target ty
 	for {
 		select {
 		case event := <-eventWatcher.ResultChan():
-			obj := event.Object.(client.Object)
-			if obj.GetName() != target.Name || obj.GetNamespace() != target.Namespace {
-				continue
-			}
-			cond, err := condition(obj)
-			if err != nil {
-				return err
-			}
-			if cond {
-				return nil
+			if event.Object != nil {
+				if obj, ok := event.Object.(client.Object); !ok {
+					continue
+				} else {
+					if obj.GetName() != target.Name || obj.GetNamespace() != target.Namespace {
+						continue
+					}
+					cond, err := condition(obj)
+					if err != nil {
+						return err
+					}
+					if cond {
+						return nil
+					}
+				}
 			}
 		case <-ctx.Done():
 			return ctx.Err()

--- a/pkg/commands/workload.go
+++ b/pkg/commands/workload.go
@@ -36,6 +36,7 @@ import (
 	corev1 "k8s.io/api/core/v1"
 	apierrs "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/types"
 
 	"github.com/vmware-tanzu/carvel-imgpkg/pkg/imgpkg/registry"
@@ -43,6 +44,7 @@ import (
 	"github.com/vmware-tanzu/apps-cli-plugin/pkg/apis"
 	cartov1alpha1 "github.com/vmware-tanzu/apps-cli-plugin/pkg/apis/cartographer/v1alpha1"
 	cli "github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime"
+	"github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime/logs"
 	"github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime/parsers"
 	cliprinter "github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime/printer"
 	"github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime/validation"
@@ -790,16 +792,28 @@ func isUrl(str string) (bool, error) {
 	}
 }
 
-func (opts *WorkloadOptions) WaitToBeReady(c *cli.Config, workload *cartov1alpha1.Workload) []wait.Worker {
-	workers := []wait.Worker{
-		func(ctx context.Context) error {
-			clientWithWatch, err := watch.GetWatcher(ctx, c)
-			if err != nil {
-				return err
-			}
-			return wait.UntilCondition(ctx, clientWithWatch, types.NamespacedName{Name: workload.Name, Namespace: workload.Namespace}, &cartov1alpha1.WorkloadList{}, cartov1alpha1.WorkloadReadyConditionFunc)
-		},
-	}
+func (opts *WorkloadOptions) GetReadyConditionWorkers(c *cli.Config, workload *cartov1alpha1.Workload, workers []wait.Worker) []wait.Worker {
+	workers = append(workers, func(ctx context.Context) error {
+		clientWithWatch, err := watch.GetWatcher(ctx, c)
+		if err != nil {
+			return err
+		}
+		return wait.UntilCondition(ctx, clientWithWatch, types.NamespacedName{Name: workload.Name, Namespace: workload.Namespace}, &cartov1alpha1.WorkloadList{}, cartov1alpha1.WorkloadReadyConditionFunc)
+	})
+
+	return workers
+}
+
+func (opts *WorkloadOptions) GetTailWorkers(c *cli.Config, workload *cartov1alpha1.Workload, workers []wait.Worker) []wait.Worker {
+	workers = append(workers, func(ctx context.Context) error {
+		selector, err := labels.Parse(fmt.Sprintf("%s=%s", cartov1alpha1.WorkloadLabelName, workload.Name))
+		if err != nil {
+			return err
+		}
+		containers := []string{}
+		return logs.Tail(ctx, c, opts.Namespace, selector, containers, time.Minute, opts.TailTimestamps)
+	})
+
 	return workers
 }
 

--- a/pkg/commands/workload_apply.go
+++ b/pkg/commands/workload_apply.go
@@ -20,16 +20,13 @@ import (
 	"context"
 	"fmt"
 	"strings"
-	"time"
 
 	"github.com/spf13/cobra"
 	apierrs "k8s.io/apimachinery/pkg/api/errors"
-	"k8s.io/apimachinery/pkg/labels"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	cartov1alpha1 "github.com/vmware-tanzu/apps-cli-plugin/pkg/apis/cartographer/v1alpha1"
 	cli "github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime"
-	"github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime/logs"
 	"github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime/validation"
 	"github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime/wait"
 	"github.com/vmware-tanzu/apps-cli-plugin/pkg/completion"
@@ -202,20 +199,14 @@ func (opts *WorkloadApplyOptions) Exec(ctx context.Context, c *cli.Config) error
 
 	if okToApply {
 		anyTail := opts.Tail || opts.TailTimestamps
+		var workers []wait.Worker
 		if opts.Wait || anyTail {
 			cli.PrintPrompt(shouldPrint, c.Infof, "Waiting for workload %q to become ready...\n", opts.Name)
 
-			workers := opts.WaitToBeReady(c, workload)
+			workers = opts.GetReadyConditionWorkers(c, workload, workers)
 
 			if anyTail {
-				workers = append(workers, func(ctx context.Context) error {
-					selector, err := labels.Parse(fmt.Sprintf("%s=%s", cartov1alpha1.WorkloadLabelName, workload.Name))
-					if err != nil {
-						return err
-					}
-					containers := []string{}
-					return logs.Tail(ctx, c, opts.Namespace, selector, containers, time.Minute, opts.TailTimestamps)
-				})
+				workers = opts.GetTailWorkers(c, workload, workers)
 			}
 
 			err := wait.Race(ctx, opts.WaitTimeout, workers)

--- a/pkg/commands/workload_create.go
+++ b/pkg/commands/workload_create.go
@@ -21,16 +21,13 @@ import (
 	"errors"
 	"fmt"
 	"strings"
-	"time"
 
 	"github.com/spf13/cobra"
 	apierrs "k8s.io/apimachinery/pkg/api/errors"
-	"k8s.io/apimachinery/pkg/labels"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	cartov1alpha1 "github.com/vmware-tanzu/apps-cli-plugin/pkg/apis/cartographer/v1alpha1"
 	cli "github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime"
-	"github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime/logs"
 	"github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime/validation"
 	"github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime/wait"
 	"github.com/vmware-tanzu/apps-cli-plugin/pkg/flags"
@@ -134,20 +131,14 @@ func (opts *WorkloadCreateOptions) Exec(ctx context.Context, c *cli.Config) erro
 
 	if okToCreate {
 		anyTail := opts.Tail || opts.TailTimestamps
+		var workers []wait.Worker
 		if opts.Wait || anyTail {
 			cli.PrintPrompt(shouldPrint, c.Infof, "Waiting for workload %q to become ready...\n", opts.Name)
 
-			workers := opts.WaitToBeReady(c, workload)
+			workers = opts.GetReadyConditionWorkers(c, workload, workers)
 
 			if anyTail {
-				workers = append(workers, func(ctx context.Context) error {
-					selector, err := labels.Parse(fmt.Sprintf("%s=%s", cartov1alpha1.WorkloadLabelName, workload.Name))
-					if err != nil {
-						return err
-					}
-					containers := []string{}
-					return logs.Tail(ctx, c, opts.Namespace, selector, containers, time.Minute, opts.TailTimestamps)
-				})
+				workers = opts.GetTailWorkers(c, workload, workers)
 			}
 
 			err := wait.Race(ctx, opts.WaitTimeout, workers)

